### ISSUE TITLE
[Performance] Optimize Gradient Accumulation for Parameter Updates

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ProjectOdyssey#2630]
## Problem

Gradient accumulation may use inefficient element-wise addition when fused operations would be faster.

## Current Pattern

```mojo
// Accumulate gradients across mini-batches
for mini_batch in batches:
    var grad = compute_gradient(mini_batch)
    accumulated_grad = add(accumulated_grad, grad)  // Creates new tensor
```

## Proposed Optimization

In-place accumulation with vectorization:

```mojo
fn accumulate_gradient_inplace[dtype: DType](
    mut accumulated: ExTensor,
    new_grad: ExTensor
):
    """Accumulate gradient in-place (no allocation).
    
    accumulated += new_grad  (vectorized)
    """
    var size = accumulated.numel()
    var acc_ptr = accumulated._data.bitcast[Scalar[dtype]]()
    var grad_ptr = new_grad._data.bitcast[Scalar[dtype]]()
    
    @parameter
    fn accumulate[idx: Int]():
        acc_ptr[idx] += grad_ptr[idx]
    
    vectorize[simd_width, accumulate](size)

// Usage
var accumulated_grad = zeros_like(parameters)
for mini_batch in batches:
    var grad = compute_gradient(mini_batch)
    accumulate_gradient_inplace(accumulated_grad, grad)  // No allocation
```

## Benefits

- No intermediate tensor allocations
- Vectorized accumulation
- Better cache utilization
- Critical for gradient accumulation in training

## Impact

- **Severity**: MEDIUM
- **Target**: Training loops with gradient accumulation
- **Benefit**: 20-30% faster gradient updates

## Related Operations

- Momentum accumulation in optimizers
- Running statistics in batch norm
- Any in-place tensor updates

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Now I have a comprehensive understanding of the codebase. Let me create the detailed implementation plan:

---

# Implementation Plan: Optimize Gradient Accumulation for Parameter Updates (Issue #2630)

## Summary

This implementation plan addresses the performance issue with gradient accumulation in the ML Odyssey codebase. Currently, gradient accumulation in `shared/autograd/tape.mojo` uses inefficient element-wise scalar loops, creating performance bottlenecks during training. The optimization involves:

1. **Creating new in-place SIMD operations** in `shared/core/inplace_simd.mojo`
2. **Optimizing `VariableRegistry.set_grad()`** to use vectorized in-place accumulation
3. **Enhancing `ExTensor.__iadd__()`** and other in-place operators to use true in-place SIMD
4. **Optimizing SGD momentum update** to use SIMD vectorization
5. **Adding comprehensive tests and benchmarks** to validate correctness and measure performance gains

**Expected Impact**: 20-30% faster gradient updates in training loops through:
- Zero intermediate allocations
- SIMD vectorized accumulation (4x speedup for float32, 2x for float64)
- Better cache utilization from in-place operations

---

## Step-by-Step Implementation Tasks

### Phase 1: Create In-Place SIMD Operations Module

**Task 1.1: Create `shared/core/inplace_simd.mojo`**

Create a new module containing in-place SIMD operations that modify tensors directly without allocation.

**File to Create**: `/home/mvillmow/ml-odyssey/shared/core/inplace_simd.mojo`

```mojo
"""In-place SIMD-optimized operations for ExTensor.

This module provides vectorized in-place operations that modify tensors
directly without intermediate allocations, achieving 2-8x speedup over
scalar implementations for large tensors.

Performance characteristics:
- float32: ~4x speedup on modern CPUs (AVX2/AVX-512)
- float64: ~2x speedup (half SIMD width of float32)
- No memory allocation overhead

Design:
- All operations modify the first argument in-place
- Shape/dtype compatibility checked before operation
- SIMD width selected at compile time based on dtype

Usage:
    from shared.core.inplace_simd import accumulate_inplace, scale_inplace
    
    var accumulated = zeros_like(grad)
    for batch in batches:
        var grad = compute_gradient(batch)
        accumulate_inplace(accumulated, grad)  # No allocation
"""

from algorithm import vectorize
from sys.info import simd_width_of
from .extensor import ExTensor
```

**Key Functions to Implement**:

1. `accumulate_inplace(mut accumulated: ExTensor, new_data: ExTensor)` - In-place addition
2. `scale_inplace(mut tensor: ExTensor, scalar: Float64)` - In-place scalar multiplication  
3. `subtract_inplace(mut a: ExTensor, b: ExTensor)` - In-place subtraction
4. `fused_momentum_update_inplace(mut velocity: ExTensor, grad: ExTensor, momentum: Float64, lr: Float64)` - Combined momentum+grad update

---

### Phase 2: Optimize Gradient Accumulation in Tape

**Task 2.1: Update `VariableRegistry.set_grad()` in `tape.mojo`**

**File to Modify**: `/home/mvillmow/ml-odyssey/shared/autograd/tape.mojo:310-341`

**Current Code** (lines 324-330):
```mojo
# Accumulate gradients - use typed access for float32
var existing = self.grads[id]
for i in range(size):
    var existing_val = existing._data.bitcast[Float32]()[i]
    var grad_val = grad._data.bitcast[Float32]()[i]
    existing._data.bitcast[Float32]()[i] = existing_val + grad_val
self.grads[id] = existing^
```

**Optimized Code**:
```mojo
# Import at top of file
from shared.core.inplace_simd import accumulate_inplace

# In set_grad():
if self.has_grad[id]:
    # In-place SIMD accumulation - no allocation
    var existing = self.grads[id]
    accumulate_inplace(existing, grad)
    self.grads[id] = existing^
else:
    # First gradient - still needs copy but use SIMD for speed
    var grad_copy = zeros_like(grad)
    accumulate_inplace(grad_copy, grad)  # SIMD copy
    self.grads[id] = grad_copy^
    self.has_grad[id] = True
```

**Task 2.2: Add dtype-generic support**

The current implementation hardcodes `Float32`. Update to dispatch based on `grad.dtype()`:
- Support `DType.float32` (primary, 4x SIMD speedup)
- Support `DType.float64` (2x SIMD speedup)
- Fallback to scalar for unsupported dtypes

---

### Phase 3: Optimize ExTensor In-Place Operators

**Task 3.1: Update `__iadd__` to use true in-place SIMD**

**File to Modify**: `/home/mvillmow/ml-odyssey/shared/core/extensor.mojo:2240-2257`

**Current Code** (allocates new tensor):
```mojo
fn __iadd__(mut self, other: ExTensor) raises:
    from .arithmetic import add
    var result = add(self, other)  # ALLOCATION!
    # Copy result data into self
    for i in range(self._numel):
        self._set_float64(i, result._get_float64(i))
```

**Optimized Code** (true in-place):
```mojo
fn __iadd__(mut self, other: ExTensor) raises:
    from .inplace_simd import accumulate_inplace
    accumulate_inplace(self, other)  # No allocation, SIMD
```

**Task 3.2: Update other in-place operators**

Apply same optimization to:
- `__isub__` → `subtract_inplace`
- `__imul__` → `multiply_inplace`
- `__itruediv__` → `divide_inplace`

---

### Phase 4: Optimize SGD Momentum Update

**Task 4.1: Add SIMD to `sgd_momentum_update_inplace`**

**File to Modify**: `/home/mvillmow/ml-odyssey/shared/training/optimizers/sgd.mojo:161-247`

**Current Code** (scalar loop, lines 225-232):
```mojo
for i in range(numel):
    velocity_data[i] = momentum_f32 * velocity_data[i] - lr_f32 * grad_data[i]
    param_data[i] += velocity_data[i]
```

**Optimized Code** (SIMD vectorized):
```mojo
alias simd_width = simd_width_of[DType.float32]()
var momentum_vec = SIMD[DType.float32, simd_width](momentum_f32)
var lr_vec = SIMD[DType.float32, simd_width](lr_f32)

@parameter
fn vectorized_update[width: Int](idx: Int) unified {mut}:
    var vel = velocity_data.load[width=width](idx)
    var grad = grad_data.load[width=width](idx)
    var param = param_data.load[width=width](idx)
    
    # velocity = momentum * velocity - lr * grad
    var new_vel = SIMD[DType.float32, width](momentum_f32) * vel - SIMD[DType.float32, width](lr_f32) * grad
    velocity_data.store[width=width](idx, new_vel)
    
    # param = param + velocity
    param_data.store[width=width](idx, param + new_vel)

vectorize[simd_width](numel, vectorized_update)
```

---

### Phase 5: Update Module Exports

**Task 5.1: Update `shared/core/__init__.mojo`**

Add exports for new in-place SIMD functions:
```mojo
from .inplace_simd import (
    accumulate_inplace,
    scale_inplace,
    subtract_inplace,
    multiply_inplace,
    divide_inplace,
)
```

---

### Phase 6: Testing

**Task 6.1: Create unit tests for in-place SIMD operations**

**File to Create**: `/home/mvillmow/ml-odyssey/tests/shared/core/test_inplace_simd.mojo`

Test cases:
1. `test_accumulate_inplace_float32` - Verify correct accumulation with float32
2. `test_accumulate_inplace_float64` - Verify correct accumulation with float64
3. `test_accumulate_inplace_shape_mismatch` - Verify error on shape mismatch
4. `test_accumulate_inplace_large_tensor` - Stress test with 1M+ elements
5. `test_scale_inplace_correctness` - Verify scalar multiplication
6. `test_subtract_inplace_correctness` - Verify subtraction
7. `test_multiple_accumulations` - Verify accumulation over many iterations

**Task 6.2: Create tests for optimized gradient accumulation**

**File to Create**: `/home/mvillmow/ml-odyssey/tests/shared/autograd/test_gradient_accumulation.mojo`

Test cases:
1. `test_set_grad_single_accumulation` - Single gradient accumulation
2. `test_set_grad_multiple_accumulations` - Multiple accumulations same variable
3. `test_set_grad_dtype_float32` - Float32 specific behavior
4. `test_set_grad_dtype_float64` - Float64 specific behavior
5. `test_set_grad_numerical_accuracy` - Compare with reference implementation
6. `test_clear_grad_after_accumulation` - Verify clear() works after accumulation

**Task 6.3: Create tests for optimized SGD**

**File to Update**: `/home/mvillmow/ml-odyssey/tests/shared/autograd/test_sgd_momentum.mojo`

Add new test cases:
1. `test_sgd_momentum_simd_correctness` - Verify SIMD gives same results as scalar
2. `test_sgd_momentum_large_params` - Test with large parameter tensors
3. `test_sgd_momentum_numerical_stability` - Test with edge values

---

### Phase 7: Benchmarking

**Task 7.1: Create performance benchmarks**

**File to Create**: `/home/mvillmow/ml-odyssey/examples/performance/benchmark_gradient_accumulation.mojo`

Benchmarks to include:
1. **Scalar vs SIMD accumulation** - Compare old vs new implementation
2. **Various tensor sizes** - Small (1K), Medium (100K), Large (10M) elements
3. **Float32 vs Float64** - Compare speedup for each dtype
4. **End-to-end training step** - Measure real-world impact

Expected output format:
```
Gradient Accumulation Benchmark
================================
Operation           Size      Scalar (ms)   SIMD (ms)    Speedup
accumulate_inplace  1K        0.12          0.04         3.0x
accumulate_inplace  100K      12.5          3.2          3.9x
accumulate_inplace  10M       1250          312          4.0x
sgd_momentum_update 100K      25.3          8.1          3.1x
```

---

## Files to Modify/Create

| File | Action | Description |
|------|--------|-------------|
| `shared/core/inplace_simd.mojo` | **Create** | New module with in-place SIMD operations |
| `shared/core/__init__.mojo` | Modify | Export new in-place SIMD functions |
| `shared/autograd/tape.mojo` | Modify | Optimize `set_grad()` with SIMD accumulation |
| `shared/core/extensor.mojo` | Modify | True in-place `__iadd__`, `__isub__`, etc. |
| `shared/training/optimizers/sgd.mojo` | Modify | SIMD-optimized momentum update |
| `tests/shared/core/test_inplace_simd.mojo` | **Create** | Unit tests for in-place SIMD ops |
| `tests/shared/autograd/test_gradient_accumulation.mojo` | **Create** | Tests for optimized gradient accumulation |
| `tests/shared/autograd/test_sgd_momentum.mojo` | Modify | Additional SIMD correctness tests |
| `examples/performance/benchmark_gradient_accumulation.mojo` | **Create** | Performance benchmarks |

---

## Testing Approach

### Unit Testing Strategy

1. **Correctness Tests**: Verify SIMD implementations produce identical results to scalar
   - Use exact FP-representable test values (0.0, 0.5, 1.0, 1.5, -1.0)
   - Compare element-by-element with tolerance 1e-6
   
2. **Edge Cases**:
   - Empty tensors (numel = 0)
   - Single element tensors
   - Non-contiguous memory (views)
   - Dtype mismatches (should error)
   - Shape mismatches (should error)

3. **Numerical Stability**:
   - Large magnitude values (±1e6)
   - Small magnitude values (±1e-6)
   - Mixed positive/negative gradients

### Integration Testing

1. **Gradient Tape Integration**:
   - Full forward-backward pass with SIMD accumulation
   - Multiple variables with different shapes
   - Clear and re-accumulate cycles

2. **Optimizer Integration**:
   - Multi-step SGD with momentum
   - Verify parameter convergence unchanged
   - Compare final weights with scalar baseline

### Performance Validation

1. **Benchmark Criteria**:
   - Target: ≥3x speedup for float32 large tensors
   - Target: ≥1.5x speedup for float64 large tensors
   - No regression for small tensors

2. **Memory Validation**:
   - Profile allocation count (should be 0 for accumulation)
   - Peak memory unchanged or reduced

---

## Success Criteria

1. **Functional Correctness**:
   - [ ] All existing tests pass unchanged
   - [ ] New unit tests pass (≥95% coverage for new code)
   - [ ] Numerical accuracy within 1e-6 of scalar baseline

2. **Performance Targets**:
   - [ ] Gradient accumulation: ≥20% speedup (measured via benchmark)
   - [ ] SGD momentum update: ≥20% speedup  
   - [ ] Zero intermediate allocations in hot path

3. **Code Quality**:
   - [ ] Follows existing SIMD patterns in `activation_simd.mojo`/`arithmetic_simd.mojo`
   - [ ] Comprehensive docstrings with performance notes
   - [ ] Pre-commit hooks pass (mojo format, markdownlint)

4. **CI Integration**:
   - [ ] All CI checks pass
   - [ ] Benchmarks added to performance tracking (optional)

---

## Implementation Order

```
Week 1:
├── Task 1.1: Create inplace_simd.mojo with accumulate_inplace
├── Task 6.1: Create test_inplace_simd.mojo
└── Task 2.1-2.2: Optimize set_grad() in tape.mojo

Week 2:
├── Task 3.1-3.2: Optimize ExTensor in-place operators
├── Task 4.1: SIMD-optimize SGD momentum update
└── Task 5.1: Update module exports

Week 3:
├── Task 6.2-6.3: Complete test coverage
├── Task 7.1: Create benchmarks
└── Final validation and PR
```

---

## Risk Mitigation

| Risk | Mitigation |
|------|------------|
| SIMD gives different results | Compare with scalar baseline in tests; use exact FP values |
| Breaking existing code | Run full test suite before/after each change |
| Memory safety issues | Use existing `UnsafePointer` patterns from `arithmetic_simd.mojo` |
| Performance regression on small tensors | Benchmark across tensor sizes; add fallback if needed |
| Dtype handling complexity | Start with float32/float64 only; document limitations |

---

## Dependencies

- **No external dependencies** - uses existing Mojo stdlib (`algorithm.vectorize`, `sys.info.simd_width_of`)
- **Internal dependencies**:
  - `shared.core.extensor.ExTensor` - tensor struct
  - `shared.autograd.tape.VariableRegistry` - gradient storage
  - `shared.training.optimizers.sgd` - optimizer implementation

---

## Resource Usage
- Tool calls used: 18/50
- Steps used: 18/50
